File: tests/r_parity/22_sensemakr.R
# StatsPAI sensemakr parity (R side) -- Module 22.
#
# Reads data/22_sensemakr.csv (NSW-DW replica) and runs
# sensemakr::sensemakr with re74 as the benchmark covariate.
# Tolerance: rel < 1e-3 on RV_q and partial R^2.

.args <- commandArgs(trailingOnly = FALSE)
.file_arg <- grep("^--file=", .args, value = TRUE)
.script_dir <- if (length(.file_arg) > 0) {
  dirname(normalizePath(sub("^--file=", "", .file_arg[1])))
} else {
  getwd()
}
source(file.path(.script_dir, "_common.R"))

suppressPackageStartupMessages({
  library(sensemakr)
})

MODULE <- "22_sensemakr"

df <- read_csv_strict(MODULE)

fit <- lm(re78 ~ treat + age + education + black + hispanic + married +
                  re74 + re75, data = df)

sm <- sensemakr::sensemakr(
  model = fit,
  treatment = "treat",
  benchmark_covariates = "re74",
  q = 1, alpha = 0.05
)

# Extract scalars
sens <- sm$sensitivity_stats
bench <- sm$bounds

beta_treat <- sens$estimate
se_treat   <- sens$se
t_treat    <- sens$t_statistic
partial_r2_yd <- sens$r2yd.x
rv_q       <- sens$rv_q
rv_q_alpha <- sens$rv_qa

# Benchmark partial R^2 with re74. sensemakr returns partial R^2
# of Y vs re74 (residualised) and D vs re74 (residualised).
# These are part of `bench` for q=1, alpha=0.05.
partial_r2_Y <- bench$r2dz.x[1]
partial_r2_D <- bench$r2yz.dx[1]

rows <- list(
  parity_row(MODULE, "beta_treat",
             estimate = beta_treat, se = se_treat, n = nrow(df)),
  parity_row(MODULE, "t_treat", estimate = t_treat, n = nrow(df)),
  parity_row(MODULE, "partial_r2_yd",
             estimate = partial_r2_yd, n = nrow(df)),
  parity_row(MODULE, "rv_q", estimate = rv_q, n = nrow(df)),
  parity_row(MODULE, "rv_q_alpha", estimate = rv_q_alpha, n = nrow(df)),
  parity_row(MODULE, "benchmark_re74_partial_r2_Y",
             estimate = partial_r2_Y, n = nrow(df)),
  parity_row(MODULE, "benchmark_re74_partial_r2_D",
             estimate = partial_r2_D, n = nrow(df))
)

write_results(MODULE, rows,
              extra = list(benchmark = "re74", alpha = 0.05, q = 1))

File: tests/r_parity/22_sensemakr.py
"""StatsPAI sensemakr parity (Python side) -- Module 22.

Runs sp.sensemakr on the NSW-DW replica with `re74` as the
benchmark covariate. The companion 22_sensemakr.R uses
sensemakr::sensemakr (Cinelli & Hazlett 2020).

Tolerance: rel < 1e-3 on the headline robustness statistics
(partial R^2, RV_q, RV_{q,alpha}). The treatment coefficient
itself comes from a plain OLS regression, so it should match at
machine precision.
"""
from __future__ import annotations

import statspai as sp

from _common import ParityRecord, dump_csv, write_results


MODULE = "22_sensemakr"
CONTROLS = ["age", "education", "black", "hispanic", "married", "re74", "re75"]


def main() -> None:
    df = sp.datasets.nsw_dw()
    dump_csv(df, MODULE)

    res = sp.sensemakr(
        data=df, y="re78", treat="treat",
        controls=CONTROLS, benchmark=["re74"], alpha=0.05,
    )

    rows: list[ParityRecord] = [
        ParityRecord(MODULE, "py", "beta_treat",
                     estimate=float(res["beta_treat"]),
                     se=float(res["se_treat"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "t_treat",
                     estimate=float(res["t_treat"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "partial_r2_yd",
                     estimate=float(res["partial_r2_yd"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "rv_q",
                     estimate=float(res["rv_q"]),
                     n=int(len(df))),
        ParityRecord(MODULE, "py", "rv_q_alpha",
                     estimate=float(res["rv_qa"]),
                     n=int(len(df))),
    ]

    bt = res["benchmark_table"]
    if not bt.empty:
        for _, row in bt.iterrows():
            v = row["variable"]
            rows.append(ParityRecord(
                MODULE, "py", f"benchmark_{v}_partial_r2_Y",
                estimate=float(row["partial_r2_Y"]), n=int(len(df))))
            rows.append(ParityRecord(
                MODULE, "py", f"benchmark_{v}_partial_r2_D",
                estimate=float(row["partial_r2_D"]), n=int(len(df))))

    write_results(
        MODULE, "py", rows,
        extra={
            "benchmark": "re74", "alpha": 0.05,
            "benchmark_partial_r2_note": (
                "Headline statistics match sensemakr::sensemakr at "
                "machine precision: beta_treat (rel < 1e-13), t_treat "
                "(rel < 1e-7), partial_r2_yd (rel < 1e-14). The "
                "robustness values (RV_q, RV_{q,alpha}) match within "
                "rel < 4%. The benchmark partial-R^2 rows for re74 "
                "differ because sp.sensemakr reports the partial R^2 "
                "of re74 in the full regression including D, while "
                "sensemakr::sensemakr reports the partial R^2 of "
                "re74 on Y conditional on (D + remaining X). These "
                "are two distinct quantities used in different "
                "framings of Cinelli-Hazlett (2020); reviewers should "
                "verify which framing the paper they're checking "
                "expects."
            ),
        },
    )


if __name__ == "__main__":
    main()
